rfcs/0003-declaration-engine.md

```diff
@@ -0,0 +1,189 @@
+
+```rust
+fn add(x: u64, y: u64) -> u64 {
+    ..
+}
+```
+
+The `TypedFunctionDeclaration` for `add` is added to the declaration engine. Then given:
+
+```rust
+let foo = add(1, 2);
+```
+
+`foo` will be a `TypedExpression` with a [`TypedExpressionVariant::FunctionApplication`](https://github.com/FuelLabs/sway/blob/dcaee960df9b3466c6897d0a86ea806c9abf7edd/sway-core/src/semantic_analysis/ast_node/expression/typed_expression_variant.rs#L18), where instead of storing the entire `TypedFunctionDeclaration` in the `TypedExpressionVariant::FunctionApplication` variant, there is a reference created `Function("add", vec!(u64, u64))` which is stored [here](https://github.com/FuelLabs/sway/blob/dcaee960df9b3466c6897d0a86ea806c9abf7edd/sway-core/src/semantic_analysis/ast_node/expression/typed_expression_variant.rs#L23). The type of `foo` is some new `TypeInfo::ReturnType("add", vec!(u64, u64))`. Both the reference `Function("add", vec!(u64, u64))` and `TypeInfo::ReturnType("add", vec!(u64, u64))` are resolved during the resolution step.
+
+Importantly, introducing the declartion engine in this way means that during type checking before the resolution step, the references introduced during usages of declarations are assumptions. There is no check happening to ensure that `add` exists when creating `Function("add", vec!(u64, u64))`. The check to see if something exists happens in the resolution step. Once the _collection context_ is introduced, the concept of checking to see if things exist can happen significantly sooner, even before type checking begins at all, if we wanted to do so.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+I believe that the only drawbacks are developer time and implementation complexity.
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+I feel that this design offers the best compromise between current state and future growth. The concept of a "declaration engine" in general is very broad, which is a good thing! The design in this RFC is a good one to get us started, but I imagine that as the language progresses, it's needs and use cases for the declaration engine will expand and change. Introducing the declaration engine now gives us the opportunity to create a mechanism through which some of those changes could be introduced later down the road.
+
+That being said, from a high-level perspective, the alternative design would be something similar to what we are doing now, which is more focused on inlining things.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+Well, I'm actually not sure. Off the top of my head, at a high-level there are languages that do not inline functions and instead use function tables, for instance. In those languages, I'm really not sure if this is done via a "declaration engine" or not. This RFC will not allow us to avoid inlining functions (we are still doing it in the IR), but it is a first step.
```

Comment:
I believe we should do at least some research before diving in, given this is a multi month journey 😄  Here are some links I think may be good starting points:
1. [rustc item collection for monomorphization](https://github.com/rust-lang/rust/blob/master/compiler/rustc_monomorphize/src/collector.rs)
2. [rustc collection item context](https://github.com/rust-lang/rust/blob/master/compiler/rustc_typeck/src/collect.rs#L105)
3. maybe helpful, maybe not, as Idris is higher level...but [idris has a context it uses to resolve types](https://github.com/idris-lang/Idris2/blob/main/src/Core/Context/Context.idr#L417)
4. Maybe @tritao has thoughts here? It seems he wrote [something similar](https://github.com/tritao/CppSharp/blob/master/src/AST/ASTContext.cs) in his older project.